code/game/machinery/computer/law.dm

```diff
@@ -28,19 +42,66 @@
 
 
 /obj/machinery/computer/aiupload/attackby(obj/item/O as obj, mob/user as mob, params)
-	if(istype(O, /obj/item/aiModule))
-		if(!current)//no AI selected
-			to_chat(user, "<span class='danger'>No AI selected. Please chose a target before proceeding with upload.")
-			return
-		var/turf/T = get_turf(current)
-		if(!atoms_share_level(T, src))
-			to_chat(user, "<span class='danger'>Unable to establish a connection</span>: You're too far away from the target silicon!")
-			return
+	if(!istype(O, /obj/item/aiModule))
+		return
+	if(!current)//no AI selected
+		to_chat(user, "<span class='danger'>No AI selected. Please chose a target before proceeding with upload.")
+		return
+	var/turf/T = get_turf(current)
+	if(!atoms_share_level(T, src))
+		to_chat(user, "<span class='danger'>Unable to establish a connection</span>: You're too far away from the target silicon!")
+		return
+	if(!emagged) //non-emag law change
 		var/obj/item/aiModule/M = O
 		M.install(src)
 		return
+	else
+		if(world.time < cooldown)
+			to_chat(user, "The program seems to have frozen. It will need some time to process.")
+			return
+		do_sparks(5, TRUE, src)
+		countlaws()
+		if(!emag_ion_check())
+			emag_inherent_law()
+		return
 	return ..()
 
+/obj/machinery/computer/aiupload/proc/countlaws()
+	foundlaws = 0
+	for(var/datum/ai_law/law in current.laws.inherent_laws)
+		foundlaws++
+	return
+
+/obj/machinery/computer/aiupload/proc/emag_ion_check()
+	if(!length(current.laws.ion_laws))
+		if(prob(20))  // 20% chance to generate an ion law if none exists
+			current.add_ion_law(generate_ion_law())
+			cooldown = world.time + EMAG_COOLDOWN
+			return TRUE
+	else //10% chance to overwrite a current ion
+		if(prob(10))
+			current.laws.ion_laws[1].law = new/datum/ai_law/ion(generate_ion_law()).law
+			cooldown = world.time + EMAG_COOLDOWN
+			return TRUE
+		else
+			return FALSE
+
+/obj/machinery/computer/aiupload/proc/emag_inherent_law()
+	if(foundlaws) //as long as it finds a law to change
+		var/lawposition = rand(1, foundlaws)
+		current.laws.inherent_laws[lawposition].law = new/datum/ai_law/inherent(generate_ion_law()).law
```

Comment:
```suggestion
		current.laws.inherent_laws[lawposition].law = new /datum/ai_law/inherent(generate_ion_law()).law
```
I would also prefer that you try and access attributes of the new'd datum on a different line for code readability